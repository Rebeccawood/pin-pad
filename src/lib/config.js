const hexStringToRGB = (color) => {
  const result = color.match(/([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})/i);
  return `${parseInt(result[1], 16)}, ${parseInt(result[2], 16)}, ${parseInt(
    result[3],
    16
  )}`;
};

export default {
  resolutions: {
    mobileL: "500px",
  },
  palette: {
    green: "#4b5435",
    greenTransparent(alpha) {
      return `rgba(${hexStringToRGB(this.green)}, ${alpha})`;
    },
    pink: "#d90755",
    pinkTransparent(alpha) {
      return `rgba(${hexStringToRGB(this.pink)}, ${alpha})`;
    },
    white: "#ffffff",
    whiteTransparent(alpha) {
      return `rgba(${hexStringToRGB(this.white)}, ${alpha})`;
    },
    offWhitePrimary: "#e8dee6",
    offWhiteSecondary: "#e8dde5",
    grey: "#7a7d7b",
  },

  borderRadius: {
    desktop: "2rem",
    mobile: "1rem",
  },
  border: {
    green: "0.1rem solid #4b5435",
    white: "0.1rem solid white",
  },
};
