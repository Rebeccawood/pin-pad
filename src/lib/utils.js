const hideNumbers = (digits) => {
  const displayNumberString = digits.join("");
  const masked = displayNumberString.substring(
    0,
    displayNumberString.length - 1
  );
  const lastNum = displayNumberString.substr(displayNumberString.length - 1);
  const hiddenNumbers = masked.replace(/./g, "*");
  return hiddenNumbers.concat(lastNum);
};

const buttonDigits = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];

const checkPin = (pin, digits) =>
  JSON.stringify(pin) === JSON.stringify(digits);

const showCorrectMessage = (pinChecked) => (pinChecked ? "OK!" : "ERROR");

export { hideNumbers, buttonDigits, checkPin, showCorrectMessage };
