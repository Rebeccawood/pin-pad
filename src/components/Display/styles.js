import styled from "styled-components";
import config from "../../lib/config";

const Wrapper = styled.section`
  display: flex;
  height: 20%;
  justify-content: center;
  width: 100%;

  @media (max-width: ${config.resolutions.mobileL}) {
    height: 7rem;
    margin-top: 1rem;
    width: 100%;
  }
`;
const Background = styled.section`
  background: linear-gradient(
    32deg,
    ${config.palette.white} 54%,
    ${config.palette.offWhitePrimary} 89%,
    ${config.palette.offWhiteSecondary} 100%
  );
  border-radius: ${config.borderRadius.desktop};
  border: ${config.border.green};
  height: 100%;
  margin-top: 4rem;
  text-align: center;
  width: 80%;

  & > p {
    font-size: 5rem;
    margin-top: 4rem;

    @media (max-width: ${config.resolutions.mobileL}) {
      font-size: 4rem;
      margin-top: 1.2rem;
    }
  }

  @media (max-width: ${config.resolutions.mobileL}) {
    border-radius: ${config.borderRadius.mobile};
    margin-top: 1rem;
  }
`;

const BlockScreenError = styled.section`
  align-items: center;
  display: block;
  height: 100%;
  margin-top: 6.5rem;
  position: absolute;
  width: 15%;

  & > p {
    background-color: ${config.palette.pink};
    border-radius: ${config.borderRadius.desktop};
    border: ${config.border.green};
    color: ${config.palette.white};
    font-size: 4rem;
    padding: 2rem;

    @media (max-width: ${config.resolutions.mobileL}) {
      border-radius: ${config.borderRadius.mobile};
      font-size: 2rem;
      padding: 0.5rem;
    }
  }

  @media (max-width: ${config.resolutions.mobileL}) {
    height: 30%;
    margin-top: 2.5rem;
    width: 60%;
  }
`;

export { Background, Wrapper, BlockScreenError };
