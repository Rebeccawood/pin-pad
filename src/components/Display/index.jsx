import React, { useState } from "react";
import { Wrapper, Background, BlockScreenError } from "./styles.js";
import { hideNumbers, checkPin, showCorrectMessage } from "../../lib/utils";

const Display = (props) => {
  const { displayNumber, setDisplayNumber } = props;
  const pin = [1, 1, 1, 1]; // correct pin number can come from API here, hence hardcoded
  const [error, setError] = useState(0);
  let timeOut;
  const shortInterval = 1000;
  const longInterval = 30000;

  const resetDisplay = (message) => {
    clearTimeout(timeOut);
    const errorIncrease = message ? 0 : error + 1;
    setDisplayState([], shortInterval);
    setErrorCount(errorIncrease, shortInterval);
  };
  const setDisplayState = (display, ms) => {
    timeOut = setTimeout(() => setDisplayNumber(display), ms);
    return timeOut;
  };
  const setErrorCount = (errorIncrease, ms) => {
    timeOut = setTimeout(() => setError(errorIncrease), ms);
    return timeOut;
  };

  return (
    <Wrapper>
      <Background>
        {displayNumber && displayNumber.length < 4 && (
          <p>{hideNumbers(displayNumber)}</p>
        )}
        {displayNumber && displayNumber.length === 4 && (
          <>
            <p>{showCorrectMessage(checkPin(pin, displayNumber))}</p>
            {resetDisplay(checkPin(pin, displayNumber))}
          </>
        )}
      </Background>
      {error === 3 && setErrorCount(0, longInterval) && (
        <BlockScreenError>
          <p>LOCKED</p>
        </BlockScreenError>
      )}
    </Wrapper>
  );
};

export default Display;
