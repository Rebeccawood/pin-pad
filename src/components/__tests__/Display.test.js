import React from "react";
import Display from "../Display/index.jsx";
import "@testing-library/jest-dom";
import Adapter from "enzyme-adapter-react-16";
import { configure } from "enzyme";
import renderer from "react-test-renderer";
import { hideNumbers, checkPin, showCorrectMessage } from "../../lib/utils";

configure({ adapter: new Adapter() });

describe("Display Component", () => {
  it("matches the snapshot", () => {
    const tree = renderer.create(<Display />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("hides all numbers except last", () => {
    expect(hideNumbers([1, 2, 3, 4])).toBe("***4");
    expect(hideNumbers([1, 2, 3])).toBe("**3");
    expect(hideNumbers([1, 2])).toBe("*2");
    expect(hideNumbers([1])).toBe("1");
  });
  it("checks pin and returns true if match", () => {
    expect(checkPin([1, 2, 3, 4], [1, 2, 3, 4])).toBeTruthy();
  });
  it("checks pin and returns false if not a match", () => {
    expect(checkPin([4, 3, 2, 1], [1, 2, 3, 4])).toBeFalsy();
  });
  it("returns OK! if pin matches", () => {
    expect(showCorrectMessage(true)).toBe("OK!");
  });
  it("returns ERROR if pin does not match", () => {
    expect(showCorrectMessage(false)).toBe("ERROR");
  });
});
