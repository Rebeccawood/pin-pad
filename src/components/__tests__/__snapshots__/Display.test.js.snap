// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Component matches the snapshot 1`] = `
<section
  className="sc-AxjAm hEXMYP"
>
  <section
    className="sc-AxirZ gatVRf"
  />
</section>
`;
